File: packages/elixir/test/crawlfox_test.exs
defmodule CrawlfoxTest do
  use ExUnit.Case

  test "requires api key" do
    prev = System.get_env("CRAWLFOX_API_KEY")
    System.delete_env("CRAWLFOX_API_KEY")

    try do
      assert_raise ArgumentError, fn -> Crawlfox.new() end
    after
      if prev, do: System.put_env("CRAWLFOX_API_KEY", prev)
    end
  end

  test "scrape search batch logs and retry" do
    scrapes = :counters.new(1, [])

    http = fn method, url, headers, body ->
      cond do
        String.ends_with?(url, "/v1/scrape") ->
          n = :counters.get(scrapes, 1) + 1
          :counters.add(scrapes, 1, 1)

          if n == 1 do
            {503, Jason.encode!(%{message: "busy", retryable: true})}
          else
            assert method == "POST"
            assert headers["Authorization"] == "Bearer cfx_test"
            assert String.starts_with?(headers["User-Agent"], "crawlfox-elixir/")
            payload = Jason.decode!(body)
            assert payload["url"] == "https://example.com"
            {200, Jason.encode!(%{success: true, data: %{markdown: "# Hello"}})}
          end

        String.ends_with?(url, "/v1/search") ->
          {200, Jason.encode!(%{success: true, data: %{web: [%{url: "https://example.com"}]}})}

        String.ends_with?(url, "/v1/batch") ->
          {200,
           Jason.encode!(%{
             success: true,
             count: 1,
             results: [%{success: true, data: %{markdown: "# A"}}]
           })}

        String.ends_with?(url, "/v1/logs/log1/result") ->
          {200, Jason.encode!(%{ok: true})}

        String.ends_with?(url, "/v1/logs/log1") ->
          assert method == "GET"
          {200, Jason.encode!(%{id: "log1", status: "ok"})}

        true ->
          {404, "{}"}
      end
    end

    client =
      Crawlfox.new(
        api_key: "cfx_test",
        api_url: "https://api.test",
        retry_backoff_ms: 1,
        http: http
      )

    doc = Crawlfox.scrape(client, "https://example.com", %{"formats" => ["markdown"]})
    assert doc["markdown"] == "# Hello"
    assert :counters.get(scrapes, 1) == 2

    search = Crawlfox.search(client, "crawlfox", %{"engine" => "google"})
    assert hd(search["web"])["url"] == "https://example.com"

    batch = Crawlfox.batch(client, ["https://example.org"])
    assert hd(batch["data"])["markdown"] == "# A"

    assert Crawlfox.get_log(client, "log1")["status"] == "ok"
    assert Crawlfox.get_log_result(client, "log1")["ok"] == true
  end

  test "does not retry quota" do
    n = :counters.new(1, [])

    http = fn _m, _u, _h, _b ->
      :counters.add(n, 1, 1)
      {429, Jason.encode!(%{message: "quota", retryable: false, code: "quota"})}
    end

    client = Crawlfox.new(api_key: "cfx_test", api_url: "https://api.test", retry_backoff_ms: 1, http: http)

    err =
      assert_raise Crawlfox.Error, fn ->
        Crawlfox.scrape(client, "https://example.com")
      end

    assert err.status == 429
    assert :counters.get(n, 1) == 1
  end
end

defmodule CrawlfoxLiveTest do
  use ExUnit.Case

  @moduletag :live

  test "live scrape search batch" do
    :inets.start()
    :ssl.start()
    client = Crawlfox.new(max_retries: 1)
    doc = Crawlfox.scrape(client, "https://example.com", %{"formats" => ["markdown"]})
    assert String.length(doc["markdown"] || "") > 10
    search = Crawlfox.search(client, "crawlfox", %{"engine" => "google", "num" => 3})
    assert length(search["web"]) > 0
    batch = Crawlfox.batch(client, ["https://example.com"], %{"formats" => ["markdown"]})
    assert String.length(get_in(batch, ["data", Access.at(0), "markdown"]) || "") > 5
  end
end

File: packages/java/src/test/java/io/crawlfox/CrawlFoxTest.java
package io.crawlfox;

import com.sun.net.httpserver.HttpServer;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.*;

class CrawlFoxTest {
    private HttpServer server;

    @AfterEach
    void stop() {
        if (server != null) {
            server.stop(0);
        }
    }

    @Test
    void requiresApiKey() {
        assertThrows(IllegalArgumentException.class, () -> new CrawlFox(""));
        assertThrows(IllegalArgumentException.class, () -> new CrawlFox("   "));
    }

    @Test
    void scrapeSearchBatchLogsAndRetry() throws Exception {
        AtomicInteger scrapes = new AtomicInteger();
        server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        server.createContext("/v1/scrape", ex -> {
            int n = scrapes.incrementAndGet();
            if (n == 1) {
                write(ex, 503, "{\"message\":\"busy\",\"retryable\":true}");
                return;
            }
            assertEquals("POST", ex.getRequestMethod());
            assertEquals("Bearer cfx_test", ex.getRequestHeaders().getFirst("Authorization"));
            assertTrue(ex.getRequestHeaders().getFirst("User-Agent").startsWith("crawlfox-java/"));
            String body = new String(ex.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
            assertTrue(body.contains("example.com"));
            write(ex, 200, "{\"success\":true,\"data\":{\"markdown\":\"# Hello\"}}");
        });
        server.createContext("/v1/search", ex ->
                write(ex, 200, "{\"success\":true,\"data\":{\"web\":[{\"url\":\"https://example.com\"}]}}"));
        server.createContext("/v1/batch", ex ->
                write(ex, 200, "{\"success\":true,\"count\":1,\"results\":[{\"success\":true,\"data\":{\"markdown\":\"# A\"}}]}"));
        server.createContext("/v1/logs/log1/result", ex -> write(ex, 200, "{\"ok\":true}"));
        server.createContext("/v1/logs/log1", ex -> {
            assertEquals("GET", ex.getRequestMethod());
            write(ex, 200, "{\"id\":\"log1\",\"status\":\"ok\"}");
        });
        server.start();
        String base = "http://127.0.0.1:" + server.getAddress().getPort();
        CrawlFox app = new CrawlFox("cfx_test", base, Duration.ofSeconds(5), 2, 1, null);

        Map<String, Object> doc = app.scrape("https://example.com", Map.of("formats", List.of("markdown")));
        assertEquals("# Hello", doc.get("markdown"));
        assertEquals(2, scrapes.get());

        @SuppressWarnings("unchecked")
        List<Map<String, Object>> web = (List<Map<String, Object>>) app.search("crawlfox", Map.of("engine", "google")).get("web");
        assertEquals("https://example.com", web.get(0).get("url"));

        @SuppressWarnings("unchecked")
        List<Map<String, Object>> batch = (List<Map<String, Object>>) app.batch(List.of("https://example.org"), Map.of()).get("data");
        assertEquals("# A", batch.get(0).get("markdown"));

        assertEquals("ok", app.getLog("log1").get("status"));
        assertEquals(true, app.getLogResult("log1").get("ok"));
    }

    @Test
    void doesNotRetryQuota() throws Exception {
        AtomicInteger n = new AtomicInteger();
        server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        server.createContext("/v1/scrape", ex -> {
            n.incrementAndGet();
            write(ex, 429, "{\"message\":\"quota\",\"retryable\":false,\"code\":\"quota\"}");
        });
        server.start();
        String base = "http://127.0.0.1:" + server.getAddress().getPort();
        CrawlFox app = new CrawlFox("cfx_test", base, Duration.ofSeconds(5), 2, 1, null);
        CrawlFox.CrawlFoxException err = assertThrows(CrawlFox.CrawlFoxException.class,
                () -> app.scrape("https://example.com"));
        assertEquals(429, err.status);
        assertEquals(1, n.get());
    }

    private static void write(com.sun.net.httpserver.HttpExchange ex, int status, String body) throws java.io.IOException {
        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
        ex.getResponseHeaders().set("Content-Type", "application/json");
        ex.sendResponseHeaders(status, bytes.length);
        try (OutputStream os = ex.getResponseBody()) {
            os.write(bytes);
        }
    }
}
